cumulus/polkadot-parachain/src/command.rs

```diff
@@ -69,7 +69,9 @@ impl RuntimeResolver for PathBuf {
 			id: String,
 		}
 
-		let file = std::fs::File::open(self).expect("Failed to open file");
+		let file = std::fs::File::open(self)
+			.unwrap_or_else(|e| panic!("Failed to open chainspec '{}': {e}", self.display()));
```

Comment:
maybe?
```suggestion.expect(&format!("File '{}' shall be readable.", self.display()));
```

Reply:
Theoretically the previous one was kinda better, since the closure in `unwrap_or_else` will only trigger if there's an error, where for the `expect` one the `format!` will always run regardless of whether there was an error or not. (So you'll always have one useless allocation; not that it really matters here, but still it triggers me a little. :P)

But why do we panic here anyway? Wouldn't it be better to return a proper error? (I can understand panics for *internal* invariants, but this is accessing the filesystem, so in theory something external could suddenly delete this file, right?)

Reply:
It is user provided file, it supposed to readable.

Reply:
Yeah the lazy formatting was the reason I went with this more verbose style originally.

As for the panicking, there is really not much to do if the provided chainspec is not readable as michal says.